### ISSUE TITLE
Partial Record destructuring

### ISSUE THREAD

**Issue by rrousselGit** [dart-lang/language#3964]
## Problem:

Currently, it is not possible to use destructuring to only extract one or two variables out of a Record.
Consider a large Record:

```dart
typedef Big = ({ int field1, int field2, ...., int field8 });
```

If we want to extract `field1`+`field2`  but nothing else, we currently have to write:

```dart
void main() {
  final Big big = (field1: 42, field2: 42, field8: 42);
  final (:field1, :field2, field3: _, field4: _, ..., field8: _) = big;
}
```

That's bloody inconvenient. It's both really tedious to type, difficult to read (due to a lot of noise), and difficult to maintain.   
In particular, adding a new field in that Record requires going through **every** places where it was destructured.

## Proposal:

When destructuring is used inside variable declarations, we could make specifying any field optional.
Using the previous `Big` record, we could therefore write:

```dart
void main() {
  final Big big = (field1: 42, field2: 42, field8: 42);
  final (:field1, :field2) = big;
}
```

This is how Record types work in Typescript.

**Comment by lrhn** [comment]
It's not completely unreasonable for a pattern with a record type as matched value type, to be able to ignore some of the fields.

The hurdle is that it's not allowed for a refutable match against a non-record type, because a record pattern must ensure that it is matched against a single known record shape.
That means `case (: var x, ...)` is only allowed against some matched value types, as the only refutable pattern that can be _invalid_ because of the type being matched.

**Comment by water-mizuu** [comment]
You can also use the defined type alias to object destructure the record.

```dart
void main() {
  final Big big = (field1: 42, field2: 42, field8: 42);
  final Big(:field1, :field2) = big;
}
```

**Comment by tatumizer** [comment]
A straightforward solution would be to enable the syntax like
```dart
final (:field1, :field2, ...var rest) = big; 
final [var x, var y, ...var rest] = [1, 2, 3, 4];
```
If the "rest" is of no interest, then
```dart
final (:field1, :field2, ..._) = big;
final [var x, var y, ..._] = [1, 2, 3, 4];
```

**Comment by munificent** [comment]
Yeah, I agree this would be useful. As @tatumizer suggests, I do think it should be something you explicitly opt into, probably using the same `...` rest syntax we have for list patterns.

However, as @lrhn notes, it's challenging to incorporate this into refutable patterns. If we support:

```dart
var stuff = (a: 1, b: 2, c: 3);
var (:a, :c, ...) = stuff;
```

Then users would probably expect this to work:

```dart
var stuff = (a: 1, b: 2, c: 3);
if (stuff case (a: 1, :var c, ...)) {
  print('a is one and c is $c');
}
```

But that might in turn lead them to expect this to work:

```dart
Object stuff = (a: 1, b: 2, c: 3); // <- Note: Has type Object now.
if (stuff case (a: 1, :var c, ...)) {
  print('a is one and c is $c');
}
```

Even though this looks similar, it's a *very* different program. Because now, what is the type that `stuff` is being tested against? Up to this point, every pattern is basically "test if the value has some type X and then do other stuff". But there's no type for "any record type that has these fields".

You can't just say "a record with fields `a` and `c`" because there's no subtype relationship between that record type and a record type like `({a: int, b: int, c: int})`. In the literature, this is called "width subtyping" and we deliberately avoided it in Dart because it makes optimizations and other stuff harder.

We could say that you can use incomplete record patterns (i.e. ones with a `...`) only in contexts where there is a surrounding context type that is a record type that fills in the `...`. That would work. I just worry that it would be a confusing limitation when users tried to use `...` in a place where there isn't that context type.

In short, it's an interesting idea and would be practically useful. I just worry about the risk of user confusion.

**Comment by rrousselGit** [comment]
@lrhn @munificent IMO it feels like we're trying to mix "pattern matching" and "destructuring" here.
To me, those are separate features, as destructuring could work without patterns.

Thinking about it: Although this issue mentioned Records, the solution probably doesn't have to be unique to them.

Instead of the existing `()` used for record patterns, we could use a different expression.
For example `:{}` (this is just an example. I'm sure a better token can be found):

```dart
final :{a, b} = (a: 42, b: 21);
final :{name}, :age = Person('john', 18);
final :{length} = [1,2,3];
final :{entries} = {'key': value};
```

And this could apply to other variable definition, such as parameters:

```dart
Event(onChange: (:{value}) => print(value));
// Equivalent to:
Event(onChange: (obj) => print(obj.value));

void fn(Person :{name}) => print(name);
// Equivalent to:
void fn(Person person) => print(person.name);
```

In that scenario, we don't care about "refutable vs irrefutable pattern" as this doesn't involve patterns at all.
And we don't need `...` either, as that syntax is by nature opt-in.

**Comment by lrhn** [comment]
This does look like an Object pattern destructuring with a different syntax.

If we had a shorthand for not writing the type, strawman `.`, then
```dart
final :{length, first} = [1,2,3];
```
would just be 
```dart
final .(:length, :first) = [1,2,3];
```
Using the existing syntax. 

We're not mixing destructuring and matching. Patterns are designed as a syntax for both, and declaration patterns are _only_ for destructuring.
It allows renaming, so you can do `var .(first: begin, last: end) = ...`, a feature you'd soon have to add to the `:{}` syntax.
 
The reason to ask for `...` is that object destructuring is not a good match records, because it doesn't handle positioal properties well. 
You have to do 
```dart
  var .($1: first, $2: second) = quintuple;
```
To get the first two positional elements of a larger record, or live with the `$i` names.
It's _nicer_ to do `var (first, second, ...) = quintuple;`, because you don't have to mention the `$i`s.
The `...` is there to say that this is not the entire type of the pattern, what it would otherwise be taken as. 
This would only work on declaration patterns, or at least it won't introduce a context type (or at least not over that can be reified as anything but `Record`). It's only allowed if the matches value type is a record type that it can match.

**Comment by rrousselGit** [comment]
> We're not mixing destructuring and matching. Patterns are designed as a syntax for both, and declaration patterns are _only_ for destructuring.

Yet the problems listed were about `case (:x, ...)` and `Object obj; var (:x) = obj`.
The former isn't a declaration pattern afaik. And the latter is only a concern if we use the "pattern" bit of the syntax to have the variable declaration perform a "match".

Both concerns, to me, feels like they are raised only because we're reusing pattern matching instead of having destructuring be its own thing.

**Comment by munificent** [comment]
@nhannah, I think your problem is different enough from one this one is originally about that it's better to fork it off to a separate issue: #4180.

Let's keep this one about partial record destructuring, which is different from how inference works with object patterns.

**Comment by water-mizuu** [comment]
> Even though this looks similar, it's a _very_ different program. Because now, what is the type that `stuff` is being tested against? Up to this point, every pattern is basically "test if the value has some type X and then do other stuff". But there's no type for "any record type that has these fields".

What if we only partial record deconstruction on explicitly Record typed objects? This would encourage them to create a type alias, or use the entire field.
